[GStreamer] Duplicate WebCore log messages due to observing the document logger multiple times

philn opened the pull request:
#### d48e67ce96f458d90843afa21b9bb56bdbb964ef
<pre>
[GStreamer] Duplicate WebCore log messages due to observing the document logger multiple times
<a href="https://bugs.webkit.org/show_bug.cgi?id=275396">https://bugs.webkit.org/show_bug.cgi?id=275396</a>

Reviewed by Xabier Rodriguez-Calvar.

Because both the PeerConnection and HTMLMediaElement rely on the same logger instance (the one
created by the document), a page containing more than one PeerConnection and/or more than one Media
element would emit duplicate logs because we were adding multiple log observers to the same logger.

Also we now filter log message by their channel name before dispatching them to the GStreamer
logging system.

Thanks to Carlos Bentzen for the report.

* Source/WebCore/Modules/mediastream/gstreamer/GStreamerPeerConnectionBackend.cpp:
(WebCore::webrtcLogObserverSingleton):
(WebCore::GStreamerPeerConnectionBackend::GStreamerPeerConnectionBackend):
(WebCore::GStreamerPeerConnectionBackend::~GStreamerPeerConnectionBackend):
(WebCore::GStreamerPeerConnectionBackend::didLogMessage): Deleted.
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerPeerConnectionBackend.h:
* Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:
(WebCore::WebCoreLogObserver::didLogMessage):
* Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.h:
(WebCore::WebCoreLogObserver::addWatch):
(WebCore::WebCoreLogObserver::removeWatch):
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::mediaLogObserverSingleton):
(WebCore::MediaPlayerPrivateGStreamer::MediaPlayerPrivateGStreamer):
(WebCore::MediaPlayerPrivateGStreamer::mediaPlayerWillBeDestroyed):
(WebCore::MediaPlayerPrivateGStreamer::createGSTPlayBin):
(WebCore::MediaPlayerPrivateGStreamer::didLogMessage): Deleted.
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.h:

Canonical link: <a href="https://commits.webkit.org/280007@main">https://commits.webkit.org/280007@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1ef04ffa738f719b9143290ed095536af84e1f2a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/55495 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/34818 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/7962 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/58480 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/5926 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/57621 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/42440 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/6123 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/44690 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/4062 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/57524 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/32727 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/47825 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/25818 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/29511 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/5153 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/4070 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/51442 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/5421 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/60071 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/30649 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/5544 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/52123 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/31734 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/47895 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/51588 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12286 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/32815 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/31481 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

calvaris commented on Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.h:
```diff
@@ -291,6 +292,37 @@ GRefPtr<GstBuffer> wrapSpanData(const std::span<const uint8_t>&);
+        if (!m_totalObservers.loadRelaxed())
+            logger.addObserver(*this);
+
+        m_totalObservers.exchangeAdd(1);
```
I would do something like:
```
auto totalObservers = m_totalObservers.exchangeAdd(1);
if (!totalObservers)
     logger.addObserver(*this);
```
Otherwise you could be racing.

philn replied:
Indeed, thanks :)

calvaris commented on Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.h:
```diff
@@ -291,6 +292,37 @@ GRefPtr<GstBuffer> wrapSpanData(const std::span<const uint8_t>&);
+        m_totalObservers.exchangeSub(1);
+
+        if (!m_totalObservers.loadRelaxed())
+            logger.removeObserver(*this);
```
Also:
```
auto totalObservers = m_totalObservers.exchangeSub(1);
if (totalObservers <= 1) {
    ASSERT(totalObservers == 1);
    logger.removeObserver(*this);
}
```

calvaris replied:
Feel free to remove the extra check or add the analog one above.

philn committed:
[GStreamer] Duplicate WebCore log messages due to observing the document logger multiple times
https://bugs.webkit.org/show_bug.cgi?id=275396

Reviewed by Xabier Rodriguez-Calvar.

Because both the PeerConnection and HTMLMediaElement rely on the same logger instance (the one
created by the document), a page containing more than one PeerConnection and/or more than one Media
element would emit duplicate logs because we were adding multiple log observers to the same logger.

Also we now filter log message by their channel name before dispatching them to the GStreamer
logging system.

Thanks to Carlos Bentzen for the report.

* Source/WebCore/Modules/mediastream/gstreamer/GStreamerPeerConnectionBackend.cpp:
(WebCore::webrtcLogObserverSingleton):
(WebCore::GStreamerPeerConnectionBackend::GStreamerPeerConnectionBackend):
(WebCore::GStreamerPeerConnectionBackend::~GStreamerPeerConnectionBackend):
(WebCore::GStreamerPeerConnectionBackend::didLogMessage): Deleted.
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerPeerConnectionBackend.h:
* Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:
(WebCore::WebCoreLogObserver::didLogMessage):
* Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.h:
(WebCore::WebCoreLogObserver::addWatch):
(WebCore::WebCoreLogObserver::removeWatch):
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::mediaLogObserverSingleton):
(WebCore::MediaPlayerPrivateGStreamer::MediaPlayerPrivateGStreamer):
(WebCore::MediaPlayerPrivateGStreamer::mediaPlayerWillBeDestroyed):
(WebCore::MediaPlayerPrivateGStreamer::createGSTPlayBin):
(WebCore::MediaPlayerPrivateGStreamer::didLogMessage): Deleted.
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.h:

Canonical link: https://commits.webkit.org/280007@main
